Lib/concurrent/futures/_base.py

```diff
@@ -170,6 +170,17 @@ def _create_and_install_waiters(fs, return_when):
 
     return waiter
 
+
+def _yield_future(fs, waiter, ref_collect=()):
+    while fs:
+        with fs[0]._condition:
```

Comment:
Anyway, this part is not relate to origin issue, so i ll remove discussed code. 

Maybe in future someone encounters it and create new issue, so it can be discussed there.